### ISSUE TITLE
not support unix socks

### ISSUE THREAD

**Issue by jan-bar** [miekg/dns#1321]
I have a piece of test code：
```go
package main

import (
	"fmt"
	"net"
	"strings"
)

func main() {
	var tcp net.Conn = &net.TCPConn{}
	_, ok := tcp.(net.PacketConn)
	fmt.Println("tcp", ok)

	var unix net.Conn = &net.UnixConn{}
	_, ok = unix.(net.PacketConn)
	fmt.Println("unix", ok)

	var udp net.Conn = &net.UDPConn{}
	_, ok = udp.(net.PacketConn)
	fmt.Println("udp", ok)
}
```
result：
tcp false
unix true
udp true

I want TCP and UNIX results to be consistent，But the end result is that UDP and UNIX are consistent，The problem code is as follows
```go
// ReadMsgHeader reads a DNS message, parses and populates hdr (when hdr is not nil).
// Returns message as a byte slice to be parsed with Msg.Unpack later on.
// Note that error handling on the message body is not possible as only the header is parsed.
func (co *Conn) ReadMsgHeader(hdr *Header) ([]byte, error) {
	var (
		p   []byte
		n   int
		err error
	)

// FIXME   The problem code is as follows
	if _, ok := co.Conn.(net.PacketConn); ok {
		if co.UDPSize > MinMsgSize {
			p = make([]byte, co.UDPSize)
		} else {
			p = make([]byte, MinMsgSize)
		}
		n, err = co.Read(p)
	} else {
		var length uint16
		if err := binary.Read(co.Conn, binary.BigEndian, &length); err != nil {
			return nil, err
		}

		p = make([]byte, length)
		n, err = io.ReadFull(co.Conn, p)
	}

	if err != nil {
		return nil, err
	} else if n < headerSize {
		return nil, ErrShortRead
	}

	p = p[:n]
	if hdr != nil {
		dh, _, err := unpackMsgHdr(p, 0)
		if err != nil {
			return nil, err
		}
		*hdr = dh
	}
	return p, err
}
```
```go
// Write implements the net.Conn Write method.
func (co *Conn) Write(p []byte) (int, error) {
	if len(p) > MaxMsgSize {
		return 0, &Error{err: "message too large"}
	}

//   FIXME  The problem code is as follows
	if _, ok := co.Conn.(net.PacketConn); ok {
		return co.Conn.Write(p)
	}

	msg := make([]byte, 2+len(p))
	binary.BigEndian.PutUint16(msg, uint16(len(p)))
	copy(msg[2:], p)
	return co.Conn.Write(msg)
}
```
At present, I use the following methods in my project
```go
conn, err = net.DialTimeout(c.Net, addr, c.Timeout)
if _, ok := conn.(*net.UnixConn); ok {
	conn = &mc{c: conn}   //  Here are my changes
}
return &dns.Conn{
	UDPSize: c.UDPSize,
	Conn:    conn,
}, nil

type mc struct{ c net.Conn }
func (c *mc) Read(b []byte) (n int, err error)   { return c.c.Read(b) }
func (c *mc) Write(b []byte) (n int, err error)  { return c.c.Write(b) }
func (c *mc) Close() error                       { return c.c.Close() }
func (c *mc) LocalAddr() net.Addr                { return c.c.LocalAddr() }
func (c *mc) RemoteAddr() net.Addr               { return c.c.RemoteAddr() }
func (c *mc) SetDeadline(t time.Time) error      { return c.c.SetDeadline(t) }
func (c *mc) SetReadDeadline(t time.Time) error  { return c.c.SetReadDeadline(t) }
func (c *mc) SetWriteDeadline(t time.Time) error { return c.c.SetWriteDeadline(t) }
```

**Comment by miekg** [comment]
[ Quoting ***@***.***> in "[miekg/dns] not support unix socks ..." ]
>I want TCP and UNIX results to be consistent，But the end result is that UDP
>and UNIX are consistent，The problem code is as follows
>
>image

can't see

>image

can't see

**Comment by jan-bar** [comment]
> [ Quoting ***@***.***> in "[miekg/dns] not support unix socks ..." ]
> I want TCP and UNIX results to be consistent，But the end result is that UDP and UNIX are consistent，The problem code is as follows image
> can't see
> image
> can't see

English is not good. I can only rely on translation software. I have changed the pictures into code fragments.

**Comment by miekg** [comment]
but what is your actual problem? You can't use a unix conn? What's the error? What are you trying to do?

**Comment by tmthrgd** [comment]
I see what's going on here and have opened #1322 to fix it.

**Comment by jan-bar** [comment]
> 

I have a dns proxy server based on unix socks.  And sending and receiving data is the same as tcp.  It doesn't work when I try to connect using unix sokcs.  So I found the above problem, but I already have my own solution. I came up to see if there is a better plan.   I see that go language supports these three interfaces `net.TCPConn, net.UDPConn, net.UnixConn`, 
 I think this <https://github.com/miekg/dns> project can be supported by the best.  Thank you so much for this awesome project.
